Add a `create_timer` method to `Node` and `LifecycleNode` classes

asymingt opened the pull request:
## Problem
Right now the `Node` and `LifecycleNode` classes only provide a method for creating a wall timer, even though the underlying functions for creating a generic timer (which respects ROS time via `/clock`) are available. This is a problem because those that write node implementations -- and especially those that come from a ROS1 background -- default to using `{Node, LifecycleNode}::create_wall_timer` not understanding the implications this decision has on timer behavior in replay, and especially when `ros2 bag play --rate` is used. The existing method of creating the correct timer type is opaque: it requires knowledge of free-functions in `rclcpp`.

Relates to issues:
- https://github.com/ros2/rclcpp/issues/465
- https://github.com/ros2/rclcpp/issues/1259

Relates to pull requests:
- https://github.com/ros2/rclcpp/pull/1982

## Fix
This PR adds the piping to allow a timer that respects `/clock` when `use_sim_time: True` node parameter is set. While it is possible to do this manually with `rclcpp::create_timer` it's much less convenient. An analogous function [already exists](https://github.com/ros2/rclpy/blob/rolling/rclpy/rclpy/node.py#L1485-L1514) in `rclpy`.

## Progress
- [x] Add the `create_timer` method to `Node` class.
- [x] Take care of edge cases in `chrono::duration` casts.
- [x] Add unit tests to `create_timer`, `timer_source` and `timer`.
- [x] Debug `TestTimeSource.check_sim_time_updated_in_callback_if_use_clock_thread`.
- [x] Update`LifecycleNode` to also support `create_timer`.
- [x] Refactor `create_timer` to drop raw pointers and have common error checks.

jacobperron commented:
Make sure to also sign-off your commits to satisfy the DCO check.

jacobperron reviewed (commented):
Looks good to me. Please add a test exercising the new method, then I'll trigger CI.

asymingt commented:
> Looks good to me. Please add a test exercising the new method, then I'll trigger CI.

Done. Please take a look at the unit tests, and particularly the `EXPECT_FALSE(timer->is_steady()` for the `GENERIC_TIMER` implementation.

Unit tests pass locally:

```txt
83: [==========] Running 16 tests from 1 test suite.
83: [----------] Global test environment set-up.
83: [----------] 16 tests from PerTimerType/TestTimer
83: [ RUN      ] PerTimerType/TestTimer.test_simple_cancel/wall_timer
83: [       OK ] PerTimerType/TestTimer.test_simple_cancel/wall_timer (17 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_simple_cancel/generic_timer
83: [       OK ] PerTimerType/TestTimer.test_simple_cancel/generic_timer (5 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_is_canceled_reset/wall_timer
83: [       OK ] PerTimerType/TestTimer.test_is_canceled_reset/wall_timer (5 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_is_canceled_reset/generic_timer
83: [       OK ] PerTimerType/TestTimer.test_is_canceled_reset/generic_timer (5 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_run_cancel_executor/wall_timer
83: [       OK ] PerTimerType/TestTimer.test_run_cancel_executor/wall_timer (105 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_run_cancel_executor/generic_timer
83: [       OK ] PerTimerType/TestTimer.test_run_cancel_executor/generic_timer (105 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_run_cancel_timer/wall_timer
83: [       OK ] PerTimerType/TestTimer.test_run_cancel_timer/wall_timer (105 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_run_cancel_timer/generic_timer
83: [       OK ] PerTimerType/TestTimer.test_run_cancel_timer/generic_timer (105 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_bad_arguments/wall_timer
83: [ERROR] [1658525415.784172078] []: Failed to fini rcl clock.
83: [       OK ] PerTimerType/TestTimer.test_bad_arguments/wall_timer (5 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_bad_arguments/generic_timer
83: [ERROR] [1658525415.788818007] []: Failed to fini rcl clock.
83: [       OK ] PerTimerType/TestTimer.test_bad_arguments/generic_timer (5 ms)
83: [ RUN      ] PerTimerType/TestTimer.callback_with_timer/wall_timer
83: [       OK ] PerTimerType/TestTimer.callback_with_timer/wall_timer (7 ms)
83: [ RUN      ] PerTimerType/TestTimer.callback_with_timer/generic_timer
83: [       OK ] PerTimerType/TestTimer.callback_with_timer/generic_timer (6 ms)
83: [ RUN      ] PerTimerType/TestTimer.callback_with_period_zero/wall_timer
83: [       OK ] PerTimerType/TestTimer.callback_with_period_zero/wall_timer (5 ms)
83: [ RUN      ] PerTimerType/TestTimer.callback_with_period_zero/generic_timer
83: [       OK ] PerTimerType/TestTimer.callback_with_period_zero/generic_timer (4 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_failures_with_exceptions/wall_timer
83: [ERROR] [1658525415.815430083] [rclcpp]: Failed to clean up rcl timer handle: error not set
83: [       OK ] PerTimerType/TestTimer.test_failures_with_exceptions/wall_timer (4 ms)
83: [ RUN      ] PerTimerType/TestTimer.test_failures_with_exceptions/generic_timer
83: [ERROR] [1658525415.820125821] [rclcpp]: Failed to clean up rcl timer handle: error not set
83: [       OK ] PerTimerType/TestTimer.test_failures_with_exceptions/generic_timer (5 ms)
83: [----------] 16 tests from PerTimerType/TestTimer (493 ms total)
83: 
83: [----------] Global test environment tear-down
83: [==========] 16 tests from 1 test suite ran. (493 ms total)
83: [  PASSED  ] 16 tests.
83: -- run_test.py: return code 0    
83: -- run_test.py: inject classname prefix into gtest result file 
The following tests passed:
        test_timer

...

100% tests passed, 0 tests failed out of 1

Label Time Summary:
gtest    =   0.68 sec*proc (1 test)

Total Test time (real) =   0.68 sec
Finished <<< rclcpp [0.94s]

Summary: 1 package finished [1.76s]
```

jacobperron commented:
>  To me it seems logical that being driven by /clock might end up making the time source non-monotonic. Does this violate any assumption made by the timer class?

Correct, if we are using a ROS clock (ie. potentially relying on the /clock topic) then we can't assume time will be be monotonically increasing. I don't think this violates any assumptions of the timer class, but I can double-check.

fujitatomoya commented on rclcpp/include/rclcpp/create_timer.hpp:
```diff
@@ -132,12 +114,97 @@ create_wall_timer(
+ */
+template<typename DurationRepT, typename DurationT, typename CallbackT>
+typename rclcpp::GenericTimer<CallbackT>::SharedPtr
+create_timer(
```
I think this needs to be tested in `rclcpp/test/rclcpp/test_create_timer.cpp`?

asymingt replied:
Looks like [simply including this header](https://github.com/ros2/rclcpp/pull/1975/commits/4594e88330dbfc3ce98e81f9f10c5d02c3737742#diff-cb1652cc3b1eff7b2611c2b78644d80d73b7e7d4185649e20a88926de473af2dR26) has triggered a CI build failure, which might be beyond my ability to solve...

clalancette replied:
> Looks like [including this header](https://github.com/ros2/rclcpp/pull/1975/commits/4594e88330dbfc3ce98e81f9f10c5d02c3737742#diff-cb1652cc3b1eff7b2611c2b78644d80d73b7e7d4185649e20a88926de473af2dR26) has triggered a CI failure, which might be beyond my ability to solve. I can take a look but I might need @jacobperron to step in.

It looks unlikely to be because of including the header.  Probably something in here is causing an endless loop, or at least a very slow loop, so that it hits the timeout.

asymingt replied:
Oh, that's interesting. Thanks, @clalancette.  I had completely misinterpreted the `[ERROR] [1659380009.202064912] [rclcpp]: failed to remove clock` as being the source of the issue. It seems too coincidental that the test took exactly one minute and zero seconds to complete. Perhaps my additional tests pushed it over some 60s internal threshold. For what it's worth, the tests pass locally so perhaps just bumping the timeout might fix the issue.

clalancette replied:
It's a little hard to see, but if you search for `test_time_source` in https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/215/consoleFull (for example), you can see it currently takes ~18 seconds.  Going beyond 60 seconds here is a huge increase, so I'd definitely say that there is something weird going on in this PR.

asymingt replied:
OK -- Now that I can see a full console log delta between builds pre- and post- my changes, it's likely that this PR is interacting negatively with simulated time in replay -- or its potentially surfacing a pre-existing issue. This agrees with local rosbag2 replay behavior I'm seeing, and might point to a larger issue with node timers driven by the `/clock` topic. I'll try and dig deeper to work out why.

asymingt replied:
@clalancette do you know if [this line of code](https://github.com/ros2/rclcpp/blob/4594e88330dbfc3ce98e81f9f10c5d02c3737742/rclcpp/test/rclcpp/test_time_source.cpp#L675) is correct? To me it looks like nanoseconds is being incremented and will overflow. Should it not rather be something like this:

```c++
pub_time_ += rclcpp::Duration(0, 1000000);
clock_msg_.clock = pub_time_;
clock_pub_->publish(clock_msg_);
```

clalancette replied:
Off-hand, I'm not sure.  That said, I think your change makes sense, so I'd give that a shot.

asymingt replied:
All failed tests happened on [agent-3af8c62f](https://build.ros2.org/computer/agent-3af8c62f). I think this CI worker needs to be spun down, as it's causing issues.

ivanpauno replied:
I commented out one test (see https://github.com/ros2/rclcpp/pull/1975#discussion_r956348057), that isn't doing what it's supposed to do.
The other tests that are flaky, I was able to reproduce the failure both for this branch and for `rolling`, so the issue is unrelated to this work.

fujitatomoya reviewed (approved):
lgtm with one question.

Blast545 commented:
@ros-pull-request-builder retest this please

asymingt commented on rclcpp/test/rclcpp/test_time_source.cpp:
```diff
@@ -695,7 +694,7 @@ class ClockThreadTestingNode : public rclcpp::Node
 
     // Create a 100ms timer
-    timer_ = this->create_wall_timer(
+    timer_ = this->create_timer(
```
We should consider reverting this, even though the reversion should be a no-op. I get that this timer sets up the periodic publishing of `/clock` to drive peer nodes. For this reason a wall timer makes more sense. However, my understanding is that since the parameter`use_sim_time` is not set on this node, it should fall back to a wall timer. Thus, both `create_timer` and `create_wall_timer` are equivalent here.

jacobperron replied:
If I understand the tests correctly, we want this to be a wall timer so I would just revert it to be safe.

ivanpauno replied:
yes, using `create_wall_timer()` here makes more sense I think

asymingt replied:
Can you double check this is what you want!

> We should consider reverting this, even though the reversion should be a no-op. I get that this timer sets up the periodic publishing of /clock to drive peer nodes' perception of time (which we call "ROS time"). For this reason a wall timer makes more sense. However, my understanding is that since the parameter use_sim_time is not set on this node, create_timer should simply fall back to a wall timer. Thus, both create_timer and create_wall_timer are equivalent here.

The above message I sent was wrong, as I incorrectly thought this change was to  the `SimClockPublisherNode` code, which it is not. If you look at how this code is used in the unit test , we have a `SimClockPublisherNode` that publishes time on `/clock` and a `ClockThreadTestingNode` that consumes this time. I'm fairly sure that we want a `create_timer` here to prove that a `/clock` driven sense of time has a callback within which you can make subsequent calls to `this->now()`. If we really wanted to test a wall timer, there would be no need to spin up a `SimClockPublisherNode` instance.

ivanpauno replied:
Yes, that seems to be the case.
I'm not sure what was the test doing before as `create_wall_timer()` never used sim time.

ivanpauno replied:
[The comments](https://github.com/ros2/rclcpp/blob/b953bdddf8de213b4a051ecf2d668dad65ff9f89/rclcpp/test/rclcpp/test_time_source.cpp#L739-L742) also say that this test should be running with `use_clock_thread=true`, and that's not the case.

When I modify the test to do what's supposed to do, it's super flaky.
IMO, it's fair to comment out the test and create a new issue for it.
@jacobperron any opinion?

ivanpauno replied:
I commented out the test as I think the problem is the test itself, see https://github.com/ros2/rclcpp/pull/1975/commits/009312185f959d93d0dde78b4a53d294aec68daa

jacobperron replied:
I thought the test was checking that the wall timer and simulation time updates were not interfering with each other. But, I'm not sure. I'm okay with disabling the flaky test. Maybe we can create a GitHub ticket to look into it (maybe it is just not a well-defined test).

ivanpauno replied:
Based on the comments, the test name, and the code; it seems that the test is supposed to be checking that the clock is updated when `use_clock_thread=True` even if you're blocked in a callback.
The part that's missing is that `use_clock_thread` was never actually enabled.
Enabling that doesn't solve the issue, as it seems something is not working well.

@anaelle-sw I think you added this feature, could you take a look at the issue?

asymingt commented:
I suspect that the nondeterminism in this PR has to do with discovery of the `/clock` topic periodically failing for some reason. If you consider the `/clock` topic is produced by some `ClockSourceNode` for consumption by `ClockSinkNode` (which has `use_sim_time: True` set as a local parameter. Internally if `ClockSinkNode` uses `this->create_timer` the resulting timer will only be advanced when `/clock` is received. However, it seems like the DDS discovery mechanism sometimes fails to find `/clock` and the timer spins indefinitely at `ros::Time(0)`. I'm not sure why this only happens on some machines and not others, but I have seen it happen on my local machine.

asymingt commented:
> @clalancette and @jacobperron -- I suspect that the nondeterminism in this PR has to do with discovery of the `/clock` topic periodically failing for some reason. If you consider the `/clock` topic is produced by some `ClockSourceNode` for consumption by `ClockSinkNode` (which has `use_sim_time: True` set as a local parameter). Internally if `ClockSinkNode` uses `this->create_timer` the resulting timer will only be advanced when `/clock` is received). However, it seems like the DDS discovery mechanism sometimes fails to find `/clock` and the timer spins indefinitely at `ros::Time(0)`. I'm not sure why this only happens on some machines and not others, but I have seen it happen on my local machine.

Update: the problem is completely fixed locally when I set `rclcpp::NodeOptions().use_clock_thread(false)` when instantiating my sink node. Perhaps there is a race condition with the clock thread implementation.

ivanpauno commented on rclcpp/include/rclcpp/create_timer.hpp:
```diff
@@ -132,12 +115,97 @@ create_wall_timer(
+  node_interfaces::NodeTimersInterface * node_timers,
+  node_interfaces::NodeClockInterface * node_clock)
+{
+  if (node_base == nullptr) {
+    throw std::invalid_argument{"input node_base cannot be null"};
+  }
+
+  if (node_timers == nullptr) {
+    throw std::invalid_argument{"input node_timers cannot be null"};
+  }
+
+  if (node_clock == nullptr) {
+    throw std::invalid_argument{"input node_clock cannot be null"};
+  }
```
can you add this checks to the `create_timer` function in L38 and reuse it?
Thanks!

It would also be nice to reorder so you first have all `create_timer()` overloads and then you have `create_wall_timer()`

asymingt replied:
This is unfortunately not trivial because it involves a function signature change. The reason is because there is a mismatch between argument types (some are raw pointers, and others are shared_ptr). What you have asked for requires a raw -> shared_ptr conversion, which is not possible.

asymingt replied:
@ivanpauno  please take a look at commit https://github.com/ros2/rclcpp/pull/1975/commits/329ec5215c39b62cb395d4644751a25b5d84b86b. It's a fairly large change that I think achieves what you have asked for. Notably it:
- Removes raw pointers in favor of shared pointers.
- Forces the create_timer call to follow a sequence of function calls.
- Moves the argument nullptr checks to a common place.
It did require making changes to some other ROS2 components to avoid passing around a raw pointer from which a shared pointer cannot be derived.

ivanpauno replied:
> This is unfortunately not trivial because it involves a function signature change. The reason is because there is a mismatch between argument types (some are raw pointers, and others are shared_ptr). If I understand your comment, what you have asked for requires a raw -> shared_ptr conversion, which is not possible.

Sorry, I didn't notice that wehn commenting, let me check your changes.

ivanpauno replied:
@asymingt sorry that my comment was confusing.
I simplified the changes in https://github.com/ros2/rclcpp/pull/1975/commits/5c129b8b17076cc9af55fc9b278ad283c6abb9b3 and https://github.com/ros2/rclcpp/pull/1975/commits/ff241d40333d31912f31570e915d0cb7665ea376

ivanpauno reviewed (commented):
I haven't looked at the tests yet, but code looks good.
I only have a minor comment.

ivanpauno commented on rclcpp/test/rclcpp/test_timer.cpp:
```diff
@@ -44,21 +51,28 @@ class TestTimer : public ::testing::Test
-      100ms,
-      [this]() -> void
-      {
+    auto lambda_fun = [this]() -> void {
```
I would use a more explicit name

```suggestion
    auto timer_callback = [this]() -> void {
```

asymingt replied:
Fixed

ivanpauno commented:
> Update: the problem is completely fixed locally when I set rclcpp::NodeOptions().use_clock_thread(false) when instantiating my ClockSinkNode. Perhaps there is a race condition with the clock thread implementation.

Could you link what's the test that was causing issues?
I went through the previous CI jobs and I see different tests failing, but I'm not sure what's the one you're referring here.

asymingt commented:
> > Update: the problem is completely fixed locally when I set rclcpp::NodeOptions().use_clock_thread(false) when instantiating my ClockSinkNode. Perhaps there is a race condition with the clock thread implementation.
> 
> Could you link what the test that was causing issues is? I went through the previous CI jobs and I see different tests failing, but I'm not sure what's the one you're referring here.

The last five commits are all empty, and in theory they should all pass. However, passing happens on only some CI workers (look at the CI build history). Locally, things pass every time. I can *occasionally* replicate the issue with test `TestTimeSource.check_sim_time_updated_in_callback_if_use_clock_thread`, but there's no single test I can show you repeatedly fails locally.

ivanpauno commented:
> The last five commits are all empty, and in theory they should all pass. However, passing happens on only some CI workers (look at the CI build history). Locally, things pass every time. I can occasionally replicate the issue with test TestTimeSource.check_sim_time_updated_in_callback_if_use_clock_thread, but there's no single test I can show you repeatedly fails locally.

Okay, that's strange.
I will take a look and see if I can reproduce locally.

ivanpauno commented:
I'm taking a look to all the failures in https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/227/#showFailuresLink and https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/228/#showFailuresLink, and they don't seem to be related to your changes at all.

It seems that for some reason they're happening in the Rpr job but not in the nightlies.

About `test_time_source`, the tests are written in a way that can easily fail.
Increasing `1s` [here](https://github.com/ros2/rclcpp/blob/b3445cec5c9bcb186c5ec8a751f05db3d6bb79c5/rclcpp/test/rclcpp/test_time_source.cpp#L77) to a longer duration may fix the issue.

asymingt commented:
> I'm taking a look to all the failures in https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/227/#showFailuresLink and https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/228/#showFailuresLink, and they don't seem to be related to your changes at all.
> 
> It seems that for some reason they're happening in the Rpr job but not in the nightlies.
> 
> About `test_time_source`, the tests are written in a way that can easily fail. Increasing `1s` [here](https://github.com/ros2/rclcpp/blob/b3445cec5c9bcb186c5ec8a751f05db3d6bb79c5/rclcpp/test/rclcpp/test_time_source.cpp#L77) to a longer duration may fix the issue.

I initially tried `10s` and that was too large, so I dropped it down to `2s` and this seems more stable.

asymingt commented:
@jacobperron and @ivanpauno - Please take a look at the last two commits and let me know if you are happy. I think I addressed all review comments, even though [this one](https://github.com/ros2/rclcpp/pull/1975#discussion_r942886593) was quite tricky. I am of the opinion that this PR is now ready to merge.

asymingt commented:
@ros-pull-request-builder retest this please

asymingt commented:
@ros-pull-request-builder retest this please

ivanpauno commented:
Something I noticed while reading the tests, this line should say `>=` and not `==`:

https://github.com/ros2/rclcpp/blob/6e0ec4a77001765a045ab9abb0505ebb36e4c43f/rclcpp/test/rclcpp/test_time_source.cpp#L84

asymingt commented:
> Something I noticed while reading the tests, this line should say `>=` and not `==`:
> 
> https://github.com/ros2/rclcpp/blob/6e0ec4a77001765a045ab9abb0505ebb36e4c43f/rclcpp/test/rclcpp/test_time_source.cpp#L84

Fixed in https://github.com/ros2/rclcpp/pull/1975/commits/c7acee7434b9cddbc9eff86dde7a7af3cfdb2958

ivanpauno reviewed (approved):
LGTM!
I will wait for @jacobperron approval as I made some minor changes.

Thanks @asymingt !!

asymingt commented:
> LGTM! I will wait for @jacobperron approval as I made some minor changes.
> 
> Thanks @asymingt !!

Thanks for getting this one over the line, Ivan 👍

jacobperron reviewed (approved):
LGTM, pending CI. One tiny nitpick.

ivanpauno commented on rclcpp/include/rclcpp/node_interfaces/node_topics.hpp:
```diff
@@ -15,6 +15,7 @@
 #ifndef RCLCPP__NODE_INTERFACES__NODE_TOPICS_HPP_
 #define RCLCPP__NODE_INTERFACES__NODE_TOPICS_HPP_
 
+#include <memory>
```
```suggestion
```

ivanpauno commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17184)](http://ci.ros2.org/job/ci_linux/17184/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11726)](http://ci.ros2.org/job/ci_linux-aarch64/11726/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=17607)](http://ci.ros2.org/job/ci_windows/17607/)

ivanpauno commented:
I forgot to include rclcpp_lifecycle and test_rclcpp, running CI again:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17187)](http://ci.ros2.org/job/ci_linux/17187/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11728)](http://ci.ros2.org/job/ci_linux-aarch64/11728/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=17609)](http://ci.ros2.org/job/ci_windows/17609/)

sloretz commented:
I'm seeing a build failure in building tf2_ros that might be related to this change: https://ci.ros2.org/job/ci_linux-aarch64/11736/console#console-section-12

```
--- stderr: tf2_ros
09:46:40 In file included from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_subscription.hpp[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/node_impl.hpp[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/node.hpp:1[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/executors/single_threaded_executor.hpp[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/executors.hpp[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/rclcpp.hpp:[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/src/ros2/geometry2/tf2_ros/include/tf2_ros/create_timer_interface.h[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/src/ros2/geometry2/tf2_ros/include/tf2_ros/create_timer_ros.h[K,
09:46:40                  from [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/src/ros2/geometry2/tf2_ros/src/create_timer_ros.cpp[K:
09:46:40 /home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp: In instantiation of ‘[Krclcpp::TimerBase::SharedPtr rclcpp::create_timer(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeTimersInterface>, rclcpp::Clock::SharedPtr, rclcpp::Duration, CallbackT&&, rclcpp::CallbackGroup::SharedPtr) [with CallbackT = std::function<void()>; rclcpp::TimerBase::SharedPtr = std::shared_ptr<rclcpp::TimerBase>; rclcpp::Clock::SharedPtr = std::shared_ptr<rclcpp::Clock>; rclcpp::CallbackGroup::SharedPtr = std::shared_ptr<rclcpp::CallbackGrou[K’:
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/src/ros2/geometry2/tf2_ros/src/create_timer_ros.cpp:61:[K   required from here
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:95:[Kerro[Kno matching function for call t[Kcreate_timer(std::__shared_ptr<rclcpp::node_interfaces::NodeBaseInterface, __gnu_cxx::_S_atomic>::element_type*, std::__shared_ptr<rclcpp::node_interfaces::NodeTimersInterface, __gnu_cxx::_S_atomic>::element_type*, rclcpp::Clock::SharedPtr&, std::chrono::duration<long int, std::ratio<1, 1000000000> >, std::function<void()>, rclcpp::CallbackGroup::SharedPt[K’
09:46:40    95 |   return [Kcreate_tim[K
09:46:40       |          [K~~~~~~~~~~[K
09:46:40    96 | [K    node_base.get[K
09:46:40       |     [K~~~~~~~~~~~~~[K  
09:46:40    97 | [K    node_timers.get[K
09:46:40       |     [K~~~~~~~~~~~~~~~[K
09:46:40    98 | [K    clo[K
09:46:40       |     [K~~~[K            
09:46:40    99 | [K    period.to_chrono<std::chrono::nanoseconds>[K
09:46:40       |     [K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
09:46:40   100 | [K    std::forward<CallbackT>(callbac[K
09:46:40       |     [K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
09:46:40   101 | [K    gro[K;
09:46:40       |     [K~~~[K            
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:87[Knot[Kcandidate[Ktemplate<class CallbackT> rclcpp::TimerBase::SharedPtr rclcpp::create_timer(std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>, std::shared_ptr<rclcpp::node_interfaces::NodeTimersInterface>, rclcpp::Clock::SharedPtr, rclcpp::Duration, CallbackT&&, rclcpp::CallbackGroup::SharedP[K’
09:46:40    87 | [Kcreate_ti[K(
09:46:40       | [K^~~~~~~~~[K
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:87[Knot[K  template argument deduction/substitution failed:
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:96:[Knot[K  cannot conver[Knode_base.std::shared_ptr<rclcpp::node_interfaces::NodeBaseInterface>::<anonymous>.std::__shared_ptr<rclcpp::node_interfaces::NodeBaseInterface, __gnu_cxx::_S_atomic>::ge[K’ (typ[Kstd::__shared_ptr<rclcpp::node_interfaces::NodeBaseInterface, __gnu_cxx::_S_atomic>::element_ty[K’ {ak[Krclcpp::node_interfaces::NodeBaseInterfa[K’}) to typ[Kstd::shared_ptr<rclcpp::node_interfaces::NodeBaseInterfa[K’
09:46:40    96 |     [Knode_base.ge[K,
09:46:40       |     [K~~~~~~~~~~~~[K
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:141[Knot[Kcandidate[Ktemplate<class DurationRepT, class DurationT, class CallbackT> typename rclcpp::GenericTimer<CallbackT>::SharedPtr rclcpp::create_timer(rclcpp::Clock::SharedPtr, std::chrono::duration<_Rep1, _Period1>, CallbackT, rclcpp::CallbackGroup::SharedPtr, rclcpp::node_interfaces::NodeBaseInterface*, rclcpp::node_interfaces::NodeTimersInterfac[K’
09:46:40   141 | [Kcreate_ti[K(
09:46:40       | [K^~~~~~~~~[K
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:141[Knot[K  template argument deduction/substitution failed:
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:95:[Knot[K  mismatched type[Kstd::chrono::duration<_Rep1, _Perio[K’ an[Kstd::__shared_ptr<rclcpp::node_interfaces::NodeTimersInterface, __gnu_cxx::_S_atomic>::element_ty[K’ {ak[Krclcpp::node_interfaces::NodeTimersInterfa[K’}
09:46:40    95 |   return [Kcreate_tim[K
09:46:40       |          [K~~~~~~~~~~[K
09:46:40    96 | [K    node_base.get[K
09:46:40       |     [K~~~~~~~~~~~~~[K  
09:46:40    97 | [K    node_timers.get[K
09:46:40       |     [K~~~~~~~~~~~~~~~[K
09:46:40    98 | [K    clo[K
09:46:40       |     [K~~~[K            
09:46:40    99 | [K    period.to_chrono<std::chrono::nanoseconds>[K
09:46:40       |     [K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
09:46:40   100 | [K    std::forward<CallbackT>(callbac[K
09:46:40       |     [K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
09:46:40   101 | [K    gro[K;
09:46:40       |     [K~~~[K            
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:107[Knot[Kcandidate[Ktemplate<class NodeT, class CallbackT> rclcpp::TimerBase::SharedPtr rclcpp::create_timer(NodeT, rclcpp::Clock::SharedPtr, rclcpp::Duration, CallbackT&&, rclcpp::CallbackGroup::SharedP[K’
09:46:40   107 | [Kcreate_ti[K(
09:46:40       | [K^~~~~~~~~[K
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:107[Knot[K  template argument deduction/substitution failed:
09:46:40 [K/home/jenkins-agent/workspace/ci_linux-aarch64/ws/install/rclcpp/include/rclcpp/rclcpp/create_timer.hpp:95:[Knot[K  candidate expects 5 arguments, 6 provided
09:46:40    95 |   return [Kcreate_tim[K
09:46:40       |          [K~~~~~~~~~~[K
09:46:40    96 | [K    node_base.get[K
09:46:40       |     [K~~~~~~~~~~~~~[K  
09:46:40    97 | [K    node_timers.get[K
09:46:40       |     [K~~~~~~~~~~~~~~~[K
09:46:40    98 | [K    clo[K
09:46:40       |     [K~~~[K            
09:46:40    99 | [K    period.to_chrono<std::chrono::nanoseconds>[K
09:46:40       |     [K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
09:46:40   100 | [K    std::forward<CallbackT>(callbac[K
09:46:40       |     [K~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~[K
09:46:40   101 | [K    gro[K;
09:46:40       |     [K~~~[K            
```

jacobperron commented:
We probably should have run with test args `--package-above rclcpp rclcpp_lifecycle`

sloretz commented:
Checking if it's this PR

build: `--packags-up-to tf2_ros`, test: `--packages-select tf2_ros` linux aarch64

With this PR: [![Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11737/badge/icon)](https://ci.ros2.org/job/ci_linux-aarch64/11737/)

WIthout this PR [repos file](https://gist.githubusercontent.com/sloretz/cdcec97436c787ac2178e6dd09a7e57f/raw/6854e82547468b2e702913a6346708271c0d2241/revert.repos) [![Build Status](https://ci.ros2.org/job/ci_linux-aarch64/11738/badge/icon)](https://ci.ros2.org/job/ci_linux-aarch64/11738/)
